### PR TITLE
feat(ci): Test that relay restarts don't break existing connected entities

### PR DESCRIPTION
**Pull request by jamilbk**
~~Highlights the issue hypothesized in #3666~~

This tests that restarting a Relay won't cause sustained downtime.

Sleeps have been removed as they shouldn't necessary -- removing them will better catch race conditions.

### CODE REVIEW

File: docker-compose.yml
Comment (jamilbk) [inline docker-compose.yml:1]: Added these to ensure that both the relay and api are started before any connected clients.

File: docker-compose.yml [hunk 1]
```diff
@@ -128,8 +128,6 @@ services:
     sysctls:
       - net.ipv6.conf.all.disable_ipv6=0
     depends_on:
-      gateway:
```
Comment (jamilbk) [inline docker-compose.yml]: Removing this allows us to write integration tests that test start and restart behavior of various services.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -203,7 +204,14 @@ jobs:
       - name: Seed database
         run: docker compose run elixir /bin/sh -c 'cd apps/domain && mix ecto.seed'
       - name: Start docker compose in the background
-        run: docker compose up -d api web client relay gateway dns.httpbin httpbin
+        run: |
+          # TODO: Order matters here, but it shouldn't. There seems to be some race
```
Comment (jamilbk) [inline .github/workflows/ci.yml]: cc @ReactorScram

File: scripts/tests/direct-curl-portal-down.sh [hunk 3]
```diff
@@ -4,10 +4,9 @@ set -e
 
 source "./scripts/tests/lib.sh"
 
+# Establish a channel
 client_curl_resource
```
Comment (thomaseizinger) [inline scripts/tests/direct-curl-portal-down.sh]:
```suggestion
client_curl_resource # Establish a channel
```

File: scripts/tests/direct-curl-portal-relay-down.sh [hunk 4]
```diff
@@ -4,10 +4,9 @@ set -e
 
 source "./scripts/tests/lib.sh"
 
+# Establish a channel
 client_curl_resource
```
Comment (thomaseizinger) [inline scripts/tests/direct-curl-portal-relay-down.sh]:
```suggestion
client_curl_resource # Establish a channel
```

File: scripts/tests/relayed-curl-portal-down.sh [hunk 5]
```diff
@@ -7,10 +7,9 @@ source "./scripts/tests/lib.sh"
 install_iptables_drop_rules
 trap remove_iptables_drop_rules EXIT # Cleanup after us
 
+# Establish a connection to the relay
 client_curl_resource
```
Comment (thomaseizinger) [inline scripts/tests/relayed-curl-portal-down.sh]:
```suggestion
client_curl_resource # Establish a channel
```

File: scripts/tests/relayed-curl-portal-restart.sh [hunk 6]
```diff
@@ -7,10 +7,6 @@ source "./scripts/tests/lib.sh"
 install_iptables_drop_rules
 trap remove_iptables_drop_rules EXIT # Cleanup after us
 
-client_curl_resource
```
Comment (thomaseizinger) [inline scripts/tests/relayed-curl-portal-restart.sh:10]: Why is this removed? Is this testing much if we just restart the portal before attempting to ping the resource?

File: scripts/tests/relayed-curl-relay-restart.sh [hunk 7]
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env bash
+
+set -e
+
+source "./scripts/tests/lib.sh"
+
+install_iptables_drop_rules
+trap remove_iptables_drop_rules EXIT # Cleanup after us
+
```
Comment (thomaseizinger) [inline scripts/tests/relayed-curl-relay-restart.sh:9]: Don't we need to _first_ ping the resource? Without pinging the resource, we won't have any state in the relay.

**Review by thomaseizinger — DISMISSED** [review]
I think the newly added test currently doesn't test anything.

File: scripts/tests/relayed-curl-portal-restart.sh (hunk 6 above)
Comment (jamilbk) [reply scripts/tests/relayed-curl-portal-restart.sh:10]: Yes, it tests that the client can open a new data channel if it's connected to a portal that just restarted. Otherwise, given the startup order, the client will create a data channel with a portal that started before it.

File: scripts/tests/direct-curl-portal-down.sh (hunk 3 above)
Comment (jamilbk) [reply scripts/tests/direct-curl-portal-down.sh]: Hm yeah, looks like my shell linter did this -- #3679 is open to lint our shell scripts. I can look into disabling this specific rule if it's a blocker.

File: scripts/tests/relayed-curl-relay-restart.sh (hunk 7 above)
Comment (jamilbk) [reply scripts/tests/relayed-curl-relay-restart.sh:9]: Hmm, shouldn't the first packet sent by curl achieve the same effect?

I was thinking these tests will catch more real-world regressions that users will see as we deploy new releases.

Comment (jamilbk) [reply scripts/tests/relayed-curl-relay-restart.sh:9]: Same thing here re: start order.

This tests that the establishment of new data channels for a client that is already connected to the portal still works after a relay restart.

We could (should) certainly test that already-established channels function properly too for relay restarts.

File: scripts/tests/relayed-curl-portal-restart.sh (hunk 6 above)
Comment (jamilbk) [reply scripts/tests/relayed-curl-portal-restart.sh:10]: Hm, but I see your point. Real-world clients will probably have some data channel already open. I'll revert.

Comment (jamilbk) [reply scripts/tests/relayed-curl-portal-restart.sh:10]: Updated to test both scenarios by restarting both with and without a binding

File: scripts/tests/relayed-curl-relay-restart.sh (hunk 7 above)
Comment (jamilbk) [reply scripts/tests/relayed-curl-relay-restart.sh:9]: Updated to test both scenarios by restarting both with and without a binding

File: scripts/tests/direct-curl-portal-relay-down.sh (hunk 4 above)
Comment (jamilbk) [reply scripts/tests/direct-curl-portal-relay-down.sh]: Fixed

File: scripts/tests/direct-curl-portal-down.sh (hunk 3 above)
Comment (jamilbk) [reply scripts/tests/direct-curl-portal-down.sh]: Nevermind, this was user error. Fixed.

**Review by jamilbk — COMMENTED** [review]
@thomaseizinger I think this is ready for another look. LMK if you still think it's not worth adding.

File: scripts/tests/relayed-curl-relay-restart.sh (hunk 7 above)
Comment (thomaseizinger) [reply scripts/tests/relayed-curl-relay-restart.sh:9]: > This tests that the establishment of new data channels for a client that is already connected to the portal still works after a relay restart.

Prior to a connection, there is no state on the client for a particular relay. We don't even know its IP. Thus I think the first relay restart doesn't do anything in regards to the client's state.

In other words, for the client, it makes no difference how many times you restart a relay before you first learn about it. It is the same state.

Restarting a relay _after_ is a good test case.

I think we should remove the first restart because it is confusing as to why it would matter.

**Review by thomaseizinger — CHANGES_REQUESTED** [review]
Overall LGTM, just one question.

File: scripts/tests/relayed-curl-relay-restart.sh (hunk 7 above)
Comment (jamilbk) [reply scripts/tests/relayed-curl-relay-restart.sh:9]: Got it -- I was under the impression the client kept a list of which Relays to use in the `init`. Updated

**Review by thomaseizinger — APPROVED** [review]
